eZ/Publish/Core/FieldType/Keyword/KeywordStorage/Gateway/LegacyStorage.php

```diff
@@ -286,12 +287,11 @@ protected function insertKeywords(array $keywordsToInsert, $contentTypeId)
         return $keywordIdMap;
     }
 
-    protected function deleteOldKeywordAssignments($fieldId, VersionInfo $versionInfo = null)
+    protected function deleteOldKeywordAssignments($fieldId, VersionInfo $versionInfo)
     {
         $dbHandler = $this->getConnection();
         // do not delete assignments if there's other published version containing this field
-        if ($versionInfo !== null &&
-            $versionInfo->status !== VersionInfo::STATUS_PUBLISHED &&
+        if ($versionInfo->status === VersionInfo::STATUS_ARCHIVED &&
```

Comment:
@andrerom there's another fix in d9c4b68. 
I'm going back to `$versionInfo` defaulting to `null`.
The reasoning behind this is that `deleteOldKeywordAssignments` is called several times during a process of adding field definitions, creating/updating field values etc., so the state of `$versionInfo` might not reflect the final "version" of this ValueObject and once we delete keywords at some point, there's no going back. 
If you want, I can dig into this more, to check why exactly we can't always pass there `$versionInfo`.

Anyway, by making this argument optional, the method behaves like it did before changes, but when actual deleting (meaning not a side effect of other operations) is called, then `$versionInfo` is passed and checked accordingly.